Feature/#5 add Asynchronous Communication message

kodachi555 opened the pull request:
WHAT
非同期通信によるメッセージ送信機能追加
(誤ってユーザのインクリメンタルサーチを途中まで進めてあります)
WHY
チャットメッセージの非同期送信のため

kodachi555 committed:
Merge branch 'develop' into feature/#5-Add-AS

kodachi555 committed:
add-text-with-input-for-group-form

kodachi555 committed:
add-users-icremental

WHAT
グループの作成、編集時にユーザ情報を
インクリメンタルで表示する
moooviの昨日の要領で作成
「追加」のリンクは未作成

kodachi555 committed:
little-correction

kodachi555 committed:
Merge branch 'feature/#5-Add-AS' into feature/Add-AS-message

kodachi555 committed:
Add-ajax-for-consolelog-formdata

kodachi555 committed:
add-messageFunction

WHAT
非同期通信でメッセージ送信の実装

kodachi555 committed:
Merge branch 'feature/#5-Add-AS-message' into feature/#5-Add-AS-Message-group

kodachi555 committed:
correction

WHAT
・メッセージ入力後はテキストフィールドを空にする
　ー メッセージが残ってしまうため
・return falseを追加
　ー 処理を完了させて、連続のメッセージ送信を可能にする
・大枠をターボリンクスが読まれた時に発火するように変更
　ー ページ遷移直後はイベントの発火が起きなくなってしまうため

kodachi555 committed:
Merge branch 'feature/#5-Add-AS-Message-group' into feature/#5-Add-AS-message

kodachi555 committed:
redo-bundle-install

MrUrakou commented on .gitignore:
```diff
@@ -15,3 +15,5 @@
 # Ignore Byebug command history file.
 .byebug_history
+
+*.png
```
pngファイル以外の画像もgitignoreしたいですね！

MrUrakou commented on config/routes.rb:
```diff
@@ -4,7 +4,12 @@
-  resources :users, only: [:edit, :update, :index]
+  resources :users, only: [:edit, :update] do
+    collection do
+      get 'index'
```
collection使用する理由ってありますか？普通にresourcesに入れるではダメですか？何らかの意図がおありで、違っていたらすみません。

kodachi555 replied:
<img width="767" alt="2018-03-26 16 11 11" src="https://user-images.githubusercontent.com/36874465/37891764-6e774932-3110-11e8-99cc-8c370e3f6540.png">
画像のようなエラーが出力されてしまうため、collectionを使用しました

MrUrakou commented on app/controllers/messages_controller.rb:
```diff
@@ -9,9 +9,12 @@ def index
-    puts @message
     if @message.save
-      redirect_to group_messages_path(@group), notice: 'メッセージが送信されました'
+      # redirect_to group_messages_path(@group), notice: 'メッセージが送信されました'
```
必要ないのであれば思い切って削除しましょう。

kodachi555 replied:
削除しました

MrUrakou commented on app/controllers/messages_controller.rb:
```diff
@@ -9,9 +9,12 @@ def index
-      redirect_to group_messages_path(@group), notice: 'メッセージが送信されました'
+      # redirect_to group_messages_path(@group), notice: 'メッセージが送信されました'
+      respond_to do |format|
+        format.html
```
jsonしか使用しないのであればhtmlの記述も必要ないですね。

kodachi555 replied:
削除しました

MrUrakou commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,42 @@
+  $('#new_message').on('submit',function(e){
+    e.preventDefault();
+    var input = new FormData(this);
+    var action = $(this).attr("action");
```
この右辺でurlは取得できてます。なのでこれをvar urlとかに代入してajaxに指定してもらえれば動くはずです。なので下の1行は必要ないです。

kodachi555 replied:
訂正しました
var url = $(this).attr("action");
    $.ajax({
      type: 'POST',
      url: url,

MrUrakou commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,42 @@
+    .done(function(datas){
+      appendMessage(datas);
+      $(".chat-main__body").animate({scrollTop:appendTarget.height()});
+      $("#message_body").val('');
```
これだと連打出来ない使用になってるはずなので、連打実装してみましょう！

kodachi555 replied:
連打の意味がよくわかりませんが、継続して投稿可能です

MrUrakou commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,42 @@
+
+  $('#new_message').on('submit',function(e){
+    e.preventDefault();
+    var input = new FormData(this);
```
FormDataをnewしてるので、var fromDataに代入しましょう。

kodachi555 replied:
修正しました

MrUrakou commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,42 @@
+    $.ajax({
+      type: 'POST',
+      url: $url,
+      data: input,
```
こちらもformDataにしましょう。

kodachi555 replied:
修正しました

MrUrakou commented on app/assets/javascripts/index.js:
```diff
@@ -1,15 +1,30 @@
+  $('#user-search-field').on('keyup',function(e){
+    e.preventDefault();
+    var input=$('#user-search-field').val();
+    $("#chat-group-users").empty();
```
こちらにinputが空の場合、ajaxが動かない記述が必要ですね。

kodachi555 replied:
if文で囲みました

MrUrakou commented on app/assets/javascripts/index.js:
```diff
@@ -1,15 +1,30 @@
     })
     .done(function(users){
-
+      if (users.length !== 0){
```
出来たらajaxを動かす前に記述したいです。

kodachi555 replied:
if文で囲みました

MrUrakou reviewed (commented):
メッセージの非同期通信とインクリメンタルサーチはブランチ分けたかったですね。あとindex.jsとかだと何のindexかぱっと見分からないのでmessage.jsにならってuser.jsにしましょう！

kodachi555 committed:
correction-UNcomment-from-messages_controller

kodachi555 committed:
correction-UN-format-html-from-messages-controller

kodachi555 committed:
correction-ajax-url

kodachi555 committed:
correction-message-js-varname-input-to-formData

kodachi555 committed:
correction-for-null-message

kodachi555 committed:
correction-for-null-input-userindex

kodachi555 commented:
ユーザのインクリメンタルサーチを途中まで進めてしまっています

kodachi555 committed:
corection-git-ignore-IgnoreUploads

kodachi555 committed:
correction-ContinuityImages

WHAT
$("#message_image").val('');
WHY
テキストが空になって、一見すると投稿後に入力内容が空になっている
ように見えたが、画像が入っていたままだったため、
空にする処理をjsファイルに追加

kodachi555 commented:
・gitignoreを訂正

・メッセージ送信後、選択した画像が残っていたため、送信後は画像も空にするように訂正

ta910 commented on config/routes.rb:
```diff
@@ -4,7 +4,12 @@
-  resources :users, only: [:edit, :update, :index]
+  resources :users, only: [:edit, :update] do
+    collection do
+      get 'index'
```
こちらcollectionでなく、resourcesに入れましょう。
jsのajaxのところでurlがusers/indexとなっているのでusers/とすれば大丈夫かと思います

ta910 commented on app/assets/javascripts/index.js:
```diff
@@ -1,15 +1,32 @@
+    if(input !== null){
+      $.ajax({
+        type: 'GET',
+        url: '/users/index',
```
こちらですね

kodachi555 committed:
correction-filename-index.js-to-user.js

kodachi555 committed:
correction-routes

kodachi555 committed:
correction-MinorFixes

kodachi555 commented:
訂正
メッセージ送信機能
・非同期通信のメッセージ、message.jsの5行目、if文を三項演算子に変更
ユーザインクリメンタルサーチ機能（途中）
・ルーティングとajaxのurl指定を訂正
